// @serialport/bindings-cpp/dist/poller.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Poller = exports.EVENTS = void 0;
const debug_1 = __importDefault(require("debug"));
const events_1 = require("events");
const path_1 = require("path");
const node_gyp_build_1 = __importDefault(require("node-gyp-build"));
const errors_1 = require("./errors");
const { Poller: PollerBindings } = (0, node_gyp_build_1.default)((0, path_1.join)(__dirname, '../'));
const logger = (0, debug_1.default)('serialport/bindings-cpp/poller');
exports.EVENTS = {
    UV_READABLE: 0b0001,
    UV_WRITABLE: 0b0010,
    UV_DISCONNECT: 0b0100,
};
function handleEvent(error, eventFlag) {
    if (error) {
        logger('error', error);
        this.emit('readable', error);
        this.emit('writable', error);
        this.emit('disconnect', error);
        return;
    }
    if (eventFlag & exports.EVENTS.UV_READABLE) {
        logger('received "readable"');
        this.emit('readable', null);
    }
    if (eventFlag & exports.EVENTS.UV_WRITABLE) {
        logger('received "writable"');
        this.emit('writable', null);
    }
    if (eventFlag & exports.EVENTS.UV_DISCONNECT) {
        logger('received "disconnect"');
        this.emit('disconnect', null);
    }
}
/**
 * Polls unix systems for readable or writable states of a file or serialport
 */
class Poller extends events_1.EventEmitter {
    constructor(fd, FDPoller = PollerBindings) {
        logger('Creating poller');
        super();
        this.poller = new FDPoller(fd, handleEvent.bind(this));
    }
    /**
     * Wait for the next event to occur
     * @param {string} event ('readable'|'writable'|'disconnect')
     * @returns {Poller} returns itself
     */
    once(event, callback) {
        switch (event) {
            case 'readable':
                this.poll(exports.EVENTS.UV_READABLE);
                break;
            case 'writable':
                this.poll(exports.EVENTS.UV_WRITABLE);
                break;
            case 'disconnect':
                this.poll(exports.EVENTS.UV_DISCONNECT);
                break;
        }
        return super.once(event, callback);
    }
    /**
     * Ask the bindings to listen for an event, it is recommend to use `.once()` for easy use
     * @param {EVENTS} eventFlag polls for an event or group of events based upon a flag.
     */
    poll(eventFlag = 0) {
        if (eventFlag & exports.EVENTS.UV_READABLE) {
            logger('Polling for "readable"');
        }
        if (eventFlag & exports.EVENTS.UV_WRITABLE) {
            logger('Polling for "writable"');
        }
        if (eventFlag & exports.EVENTS.UV_DISCONNECT) {
            logger('Polling for "disconnect"');
        }
        this.poller.poll(eventFlag);
    }
    /**
     * Stop listening for events and cancel all outstanding listening with an error
     */
    stop() {
        logger('Stopping poller');
        this.poller.stop();
        this.emitCanceled();
    }
    destroy() {
        logger('Destroying poller');
        this.poller.destroy();
        this.emitCanceled();
    }
    emitCanceled() {
        const err = new errors_1.BindingsError('Canceled', { canceled: true });
        this.emit('readable', err);
        this.emit('writable', err);
        this.emit('disconnect', err);
    }
}
exports.Poller = Poller;

// SIG // Begin signature block
// SIG // MIIn4AYJKoZIhvcNAQcCoIIn0TCCJ80CAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // y8KZDFmdV3pxcPLfaZz4iCiWa2L66/IsWTgfjdUvkySg
// SIG // gg2WMIIGFDCCA/ygAwIBAgITMwAAAvMJbkUH18AklgAA
// SIG // AAAC8zANBgkqhkiG9w0BAQsFADB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExMB4XDTIyMDYzMDE3MzgyMloX
// SIG // DTIzMDkxNTE3MzgyMlowgYgxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xMjAwBgNVBAMTKU1pY3Jvc29mdCAzcmQgUGFydHkg
// SIG // QXBwbGljYXRpb24gQ29tcG9uZW50MIIBIjANBgkqhkiG
// SIG // 9w0BAQEFAAOCAQ8AMIIBCgKCAQEAr60JgfZVHtyxUjDP
// SIG // RUyYBDZHgOnoau2MKAmwIGvyGGE6SjsgzyZGPZSTs0zx
// SIG // 2iKtFyM+MMAGaTPhAn+S8W+LO/bnu0PhYNTfx03gMBAz
// SIG // q0qvY2yU3ysXfH+C53qfvW/dUzYdT710qXUVQI7vpOVU
// SIG // 7e+aKGnaqKBDyf8Z0uq9cHXHaXk74em8sjqDUkm2eDJQ
// SIG // ZWic5DNEnkw8OBjQetCkLtWVHtfB+J0TyJe4bKDcwaB6
// SIG // jI2ziWxSH/3HSFpTInbHKtGXuibQoJfA+mPLYoFRGJN8
// SIG // f7pT30g5tuLNUK4QUssenZTneQA/pnsVFmfoIagAwwxR
// SIG // hrkGPtVkOwqCqxd91wIDAQABo4IBfjCCAXowHwYDVR0l
// SIG // BBgwFgYKKwYBBAGCN0wRAQYIKwYBBQUHAwMwHQYDVR0O
// SIG // BBYEFHpzR29bYqzU06mm+xIDW0/abcPgMFAGA1UdEQRJ
// SIG // MEekRTBDMSkwJwYDVQQLEyBNaWNyb3NvZnQgT3BlcmF0
// SIG // aW9ucyBQdWVydG8gUmljbzEWMBQGA1UEBRMNMjMxNTIy
// SIG // KzQ3MTMzNjAfBgNVHSMEGDAWgBRIbmTlUAXTgqoXNzci
// SIG // tW2oynUClTBUBgNVHR8ETTBLMEmgR6BFhkNodHRwOi8v
// SIG // d3d3Lm1pY3Jvc29mdC5jb20vcGtpb3BzL2NybC9NaWND
// SIG // b2RTaWdQQ0EyMDExXzIwMTEtMDctMDguY3JsMGEGCCsG
// SIG // AQUFBwEBBFUwUzBRBggrBgEFBQcwAoZFaHR0cDovL3d3
// SIG // dy5taWNyb3NvZnQuY29tL3BraW9wcy9jZXJ0cy9NaWND
// SIG // b2RTaWdQQ0EyMDExXzIwMTEtMDctMDguY3J0MAwGA1Ud
// SIG // EwEB/wQCMAAwDQYJKoZIhvcNAQELBQADggIBAICJJoIo
// SIG // aU2RkTKNc8lbsIRZS81hAwT8XrwfrI3IB7SEkQhAccHA
// SIG // Skulj2D5CQnfnA6N2xDHXwj4MjHwTMUNg5V2kp0RnOcD
// SIG // 1QuoWs5eeUuHOgMWnJrpEW1XvPbG/qW648WmCXuVJq+L
// SIG // XPfu2nf3tvMoq3aa0/9/CkSl3lcI1WUQFK6sn/UkaTob
// SIG // Q9q3DWUXIi/NiG/XW92886GhvctJAmkPjpY3oUPGTjuE
// SIG // Lr/GXtyRnre859Ci5l5F0m4GjUYmS53kHQery9hlty6w
// SIG // slWndAKYfXOEsr5VtBbmjFAkUFC+1HFVU+oFCSCMhxyi
// SIG // dVxnIHW5gBcQPZQ64KIG5nOk1MwVHwFhrIHqAMzUyWe2
// SIG // jmc4NGN2NtYbASgKdRsc9KMrNxgebO7eprCu9s80bC2R
// SIG // e+bfJLbqJK5HfAuSEtqf+oRdC7T5d2lj9JNVfgmD2cNA
// SIG // oBx1Xl+Xwv9Vd+dx0fKP09dOhet+vxuGLzKaptyU6ABQ
// SIG // LmhoVOWnqVvN2UUYwp2ZM1AmAjog3KA3Y3K6+iLtoeeD
// SIG // sKm+OCTVJ5Zgnlzcn2jwn6byPRtz4B4N0slyOuwOd9ro
// SIG // ZTU1l51WrmzI7mimgJUXhkE8XlwDnLs0O9T0F3L/O5Rz
// SIG // 3YBiQF8cpp0mgVlVV6fOmNtH0XnRbJE2FeGMHEbDZOB8
// SIG // MKnZx8vyVoYZ+0iVMIIHejCCBWKgAwIBAgIKYQ6Q0gAA
// SIG // AAAAAzANBgkqhkiG9w0BAQsFADCBiDELMAkGA1UEBhMC
// SIG // VVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcT
// SIG // B1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jw
// SIG // b3JhdGlvbjEyMDAGA1UEAxMpTWljcm9zb2Z0IFJvb3Qg
// SIG // Q2VydGlmaWNhdGUgQXV0aG9yaXR5IDIwMTEwHhcNMTEw
// SIG // NzA4MjA1OTA5WhcNMjYwNzA4MjEwOTA5WjB+MQswCQYD
// SIG // VQQGEwJVUzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4G
// SIG // A1UEBxMHUmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0
// SIG // IENvcnBvcmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQg
// SIG // Q29kZSBTaWduaW5nIFBDQSAyMDExMIICIjANBgkqhkiG
// SIG // 9w0BAQEFAAOCAg8AMIICCgKCAgEAq/D6chAcLq3YbqqC
// SIG // EE00uvK2WCGfQhsqa+laUKq4BjgaBEm6f8MMHt03a8YS
// SIG // 2AvwOMKZBrDIOdUBFDFC04kNeWSHfpRgJGyvnkmc6Whe
// SIG // 0t+bU7IKLMOv2akrrnoJr9eWWcpgGgXpZnboMlImEi/n
// SIG // qwhQz7NEt13YxC4Ddato88tt8zpcoRb0RrrgOGSsbmQ1
// SIG // eKagYw8t00CT+OPeBw3VXHmlSSnnDb6gE3e+lD3v++Mr
// SIG // WhAfTVYoonpy4BI6t0le2O3tQ5GD2Xuye4Yb2T6xjF3o
// SIG // iU+EGvKhL1nkkDstrjNYxbc+/jLTswM9sbKvkjh+0p2A
// SIG // LPVOVpEhNSXDOW5kf1O6nA+tGSOEy/S6A4aN91/w0FK/
// SIG // jJSHvMAhdCVfGCi2zCcoOCWYOUo2z3yxkq4cI6epZuxh
// SIG // H2rhKEmdX4jiJV3TIUs+UsS1Vz8kA/DRelsv1SPjcF0P
// SIG // UUZ3s/gA4bysAoJf28AVs70b1FVL5zmhD+kjSbwYuER8
// SIG // ReTBw3J64HLnJN+/RpnF78IcV9uDjexNSTCnq47f7Fuf
// SIG // r/zdsGbiwZeBe+3W7UvnSSmnEyimp31ngOaKYnhfsi+E
// SIG // 11ecXL93KCjx7W3DKI8sj0A3T8HhhUSJxAlMxdSlQy90
// SIG // lfdu+HggWCwTXWCVmj5PM4TasIgX3p5O9JawvEagbJjS
// SIG // 4NaIjAsCAwEAAaOCAe0wggHpMBAGCSsGAQQBgjcVAQQD
// SIG // AgEAMB0GA1UdDgQWBBRIbmTlUAXTgqoXNzcitW2oynUC
// SIG // lTAZBgkrBgEEAYI3FAIEDB4KAFMAdQBiAEMAQTALBgNV
// SIG // HQ8EBAMCAYYwDwYDVR0TAQH/BAUwAwEB/zAfBgNVHSME
// SIG // GDAWgBRyLToCMZBDuRQFTuHqp8cx0SOJNDBaBgNVHR8E
// SIG // UzBRME+gTaBLhklodHRwOi8vY3JsLm1pY3Jvc29mdC5j
// SIG // b20vcGtpL2NybC9wcm9kdWN0cy9NaWNSb29DZXJBdXQy
// SIG // MDExXzIwMTFfMDNfMjIuY3JsMF4GCCsGAQUFBwEBBFIw
// SIG // UDBOBggrBgEFBQcwAoZCaHR0cDovL3d3dy5taWNyb3Nv
// SIG // ZnQuY29tL3BraS9jZXJ0cy9NaWNSb29DZXJBdXQyMDEx
// SIG // XzIwMTFfMDNfMjIuY3J0MIGfBgNVHSAEgZcwgZQwgZEG
// SIG // CSsGAQQBgjcuAzCBgzA/BggrBgEFBQcCARYzaHR0cDov
// SIG // L3d3dy5taWNyb3NvZnQuY29tL3BraW9wcy9kb2NzL3By
// SIG // aW1hcnljcHMuaHRtMEAGCCsGAQUFBwICMDQeMiAdAEwA
// SIG // ZQBnAGEAbABfAHAAbwBsAGkAYwB5AF8AcwB0AGEAdABl
// SIG // AG0AZQBuAHQALiAdMA0GCSqGSIb3DQEBCwUAA4ICAQBn
// SIG // 8oalmOBUeRou09h0ZyKbC5YR4WOSmUKWfdJ5DJDBZV8u
// SIG // LD74w3LRbYP+vj/oCso7v0epo/Np22O/IjWll11lhJB9
// SIG // i0ZQVdgMknzSGksc8zxCi1LQsP1r4z4HLimb5j0bpdS1
// SIG // HXeUOeLpZMlEPXh6I/MTfaaQdION9MsmAkYqwooQu6Sp
// SIG // BQyb7Wj6aC6VoCo/KmtYSWMfCWluWpiW5IP0wI/zRive
// SIG // /DvQvTXvbiWu5a8n7dDd8w6vmSiXmE0OPQvyCInWH8My
// SIG // GOLwxS3OW560STkKxgrCxq2u5bLZ2xWIUUVYODJxJxp/
// SIG // sfQn+N4sOiBpmLJZiWhub6e3dMNABQamASooPoI/E01m
// SIG // C8CzTfXhj38cbxV9Rad25UAqZaPDXVJihsMdYzaXht/a
// SIG // 8/jyFqGaJ+HNpZfQ7l1jQeNbB5yHPgZ3BtEGsXUfFL5h
// SIG // YbXw3MYbBL7fQccOKO7eZS/sl/ahXJbYANahRr1Z85el
// SIG // CUtIEJmAH9AAKcWxm6U/RXceNcbSoqKfenoi+kiVH6v7
// SIG // RyOA9Z74v2u3S5fi63V4GuzqN5l5GEv/1rMjaHXmr/r8
// SIG // i+sLgOppO6/8MO0ETI7f33VtY5E90Z1WTk+/gFcioXgR
// SIG // MiF670EKsT/7qMykXcGhiJtXcVZOSEXAQsmbdlsKgEhr
// SIG // /Xmfwb1tbWrJUnMTDXpQzTGCGaIwghmeAgEBMIGVMH4x
// SIG // CzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9u
// SIG // MRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNy
// SIG // b3NvZnQgQ29ycG9yYXRpb24xKDAmBgNVBAMTH01pY3Jv
// SIG // c29mdCBDb2RlIFNpZ25pbmcgUENBIDIwMTECEzMAAALz
// SIG // CW5FB9fAJJYAAAAAAvMwDQYJYIZIAWUDBAIBBQCgga4w
// SIG // GQYJKoZIhvcNAQkDMQwGCisGAQQBgjcCAQQwHAYKKwYB
// SIG // BAGCNwIBCzEOMAwGCisGAQQBgjcCARUwLwYJKoZIhvcN
// SIG // AQkEMSIEIAGLdAkEmNLkATnAdBm4frCQZQb+EltxUGeH
// SIG // Aw+GnhzWMEIGCisGAQQBgjcCAQwxNDAyoBSAEgBNAGkA
// SIG // YwByAG8AcwBvAGYAdKEagBhodHRwOi8vd3d3Lm1pY3Jv
// SIG // c29mdC5jb20wDQYJKoZIhvcNAQEBBQAEggEAhI+3ix/Z
// SIG // VGOhCNdlVRoiyjAUWKOTtOaRdTTsdflwK1hFxSDShW3m
// SIG // EISo2YhAyiKVWiMeG0LMAQVVecr1yLZY5fSjNQW4e6r8
// SIG // R0Gj03sgcT75gUYeVD5nkKIVRzp7WuFtpKM3yVzadxPU
// SIG // P+Syka0j8oBmRI8cWk6MXfk0+/IHL5n0cBcwLBS7idBx
// SIG // e3e8OwiaLnAHOVYEKU94Q+4Y4MDOYcuNz6gF2mNeCP5g
// SIG // fbXXNDHdSbsi8g+oVXjH59k8KnzsGQocHUJ7R8V0uFgu
// SIG // FUYa8WqeeKNb1DCt4blPSWG6W+U8C2N2l/jCoRpTKZOd
// SIG // Ir8Zwkgq0hxAkMfgmTYS8AjL1aGCFywwghcoBgorBgEE
// SIG // AYI3AwMBMYIXGDCCFxQGCSqGSIb3DQEHAqCCFwUwghcB
// SIG // AgEDMQ8wDQYJYIZIAWUDBAIBBQAwggFYBgsqhkiG9w0B
// SIG // CRABBKCCAUcEggFDMIIBPwIBAQYKKwYBBAGEWQoDATAx
// SIG // MA0GCWCGSAFlAwQCAQUABCC9k1pg9Y+2BLP+BQ2hR3+F
// SIG // f3zbFvy5V+kyt4Iy7AA8TAIGY8fdepf5GBIyMDIzMDEy
// SIG // NTAxMzAyOC4wOFowBIACAfSggdikgdUwgdIxCzAJBgNV
// SIG // BAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYD
// SIG // VQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQg
// SIG // Q29ycG9yYXRpb24xLTArBgNVBAsTJE1pY3Jvc29mdCBJ
// SIG // cmVsYW5kIE9wZXJhdGlvbnMgTGltaXRlZDEmMCQGA1UE
// SIG // CxMdVGhhbGVzIFRTUyBFU046MTc5RS00QkIwLTgyNDYx
// SIG // JTAjBgNVBAMTHE1pY3Jvc29mdCBUaW1lLVN0YW1wIFNl
// SIG // cnZpY2WgghF8MIIHJzCCBQ+gAwIBAgITMwAAAbWtGt/X
// SIG // hXBtEwABAAABtTANBgkqhkiG9w0BAQsFADB8MQswCQYD
// SIG // VQQGEwJVUzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4G
// SIG // A1UEBxMHUmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0
// SIG // IENvcnBvcmF0aW9uMSYwJAYDVQQDEx1NaWNyb3NvZnQg
// SIG // VGltZS1TdGFtcCBQQ0EgMjAxMDAeFw0yMjA5MjAyMDIy
// SIG // MTFaFw0yMzEyMTQyMDIyMTFaMIHSMQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMS0wKwYDVQQLEyRNaWNyb3NvZnQgSXJlbGFu
// SIG // ZCBPcGVyYXRpb25zIExpbWl0ZWQxJjAkBgNVBAsTHVRo
// SIG // YWxlcyBUU1MgRVNOOjE3OUUtNEJCMC04MjQ2MSUwIwYD
// SIG // VQQDExxNaWNyb3NvZnQgVGltZS1TdGFtcCBTZXJ2aWNl
// SIG // MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA
// SIG // lwsKuGVegsKNiYXFwU+CSHnt2a7PfWw2yPwiW+YRlEJs
// SIG // H3ibFIiPfk/yblMp8JGantu+7Di/+3e5wWN/nbJUIMUj
// SIG // EWJnc8JMjoPmHCWsMtJOuR/1Ru4aa1RrxQtIelq098TB
// SIG // l4k7NsEE87l7qKFmy8iwGNQjkwr0bMu4BJwy7BUXiXHe
// SIG // gOSU992rfQ4xNZoxznv42TLQsc9NmcBq5WslkqVATcc8
// SIG // PSfgBLEpdG1Dp2wqNw4JrJFwJNA1bfzTScYABc5smRZB
// SIG // gsP4JiK/8CVrlocheEyQonjm3rFttrojAreSUnixALu9
// SIG // pDrsBI4DUPGG34oIbieI1oqFl/xk7A+7uM8k4o8ifMVW
// SIG // NTaczbPldDYtn6hBre7r25RED4uecCxP8Dxy34YPUElW
// SIG // llPP3LAXp5cMwRjx+EWzjEtILEKXuAcfxrXCTwyYhm5X
// SIG // NzCCZYh4/gF2U2y/bYfekKpaoFYwkoZeT6ZxoQbX5Kft
// SIG // gj+tZkFV21UvZIkJ6b34a/44dtrsK6diTmVnNTM9J6P6
// SIG // Ehlk2sfcUwbHIGL8mYqdKOiyd4RxOCmSvcFNkZEgrk54
// SIG // 8mHCbDbTyO9xSzN1EkWxbp8n/LHVnZ9fp5hILGntkMza
// SIG // D5aXRCQyHSIhsPtR7Q/rKoHyjFqgtGO9ftnxYvxzNrbK
// SIG // eMCzwmcqwMrX6Hcxe0SeKZ8CAwEAAaOCAUkwggFFMB0G
// SIG // A1UdDgQWBBRsUIbZgoZVXVXVWQX0Ok1VO2bHUzAfBgNV
// SIG // HSMEGDAWgBSfpxVdAF5iXYP05dJlpxtTNRnpcjBfBgNV
// SIG // HR8EWDBWMFSgUqBQhk5odHRwOi8vd3d3Lm1pY3Jvc29m
// SIG // dC5jb20vcGtpb3BzL2NybC9NaWNyb3NvZnQlMjBUaW1l
// SIG // LVN0YW1wJTIwUENBJTIwMjAxMCgxKS5jcmwwbAYIKwYB
// SIG // BQUHAQEEYDBeMFwGCCsGAQUFBzAChlBodHRwOi8vd3d3
// SIG // Lm1pY3Jvc29mdC5jb20vcGtpb3BzL2NlcnRzL01pY3Jv
// SIG // c29mdCUyMFRpbWUtU3RhbXAlMjBQQ0ElMjAyMDEwKDEp
// SIG // LmNydDAMBgNVHRMBAf8EAjAAMBYGA1UdJQEB/wQMMAoG
// SIG // CCsGAQUFBwMIMA4GA1UdDwEB/wQEAwIHgDANBgkqhkiG
// SIG // 9w0BAQsFAAOCAgEAkFGOpyjKV2s2sA+wTqDwDdhp0mFr
// SIG // PtiU4rN3OonTWqb85M6WH19c/P517xujLCih/HllP5xK
// SIG // WmXnAIRV1/NQDkJBLSdLTb/NQtcT1FWGQ7CMTnrn9tLZ
// SIG // xqIFtKVylvQNyh31C/qkC8QmNpyzakO0G38uOGgOkJ9E
// SIG // q4nA+7QwVfobDlggWuEpzdFnRdyXL32gOqSvrLjFKpv4
// SIG // KEVqaBTiaxCWZDlIhG3YgUza7cnG5Z2SA/feMq/IiV06
// SIG // AzUadZw6XgcTrqXmEmE0tMmdl44MMFC3wGU9AVeFCWKd
// SIG // D9WOnYA2zHg+XF2LQVto0VYtFLd6c6DQFcmB38GvPCKV
// SIG // YSn8r10EoXuRN+gQ7hLcim12esOnW4F4bHCmHWTVWeAG
// SIG // gPiSItHHRfGKLEUZmotVOdFPR8wiuADT/fHSXBkkdpL1
// SIG // 2tvgEGELeTznzFulZ16b/Nv6dtbgSRZreesJBNKpTjdY
// SIG // ju/GqnlAkpflL6J0wxk957/UVYnmjjRY61jX90QGQmBz
// SIG // m9vs/+2bj02Xx/bXXy8vq57jmNXQ2ufOaJm3nAcD2qOa
// SIG // SyXEOj9mqhMt4tdvMjHhiNPldfj0Q7Kq1HgdRBrKWkzC
// SIG // QNi4ts8HRJBipNaVpWfU7BcRn8BeYzdLoIzwRLDtatz6
// SIG // aBho3oD/bXHrZagxprM5MsMB/rVfb5Xn1YS7/uEwggdx
// SIG // MIIFWaADAgECAhMzAAAAFcXna54Cm0mZAAAAAAAVMA0G
// SIG // CSqGSIb3DQEBCwUAMIGIMQswCQYDVQQGEwJVUzETMBEG
// SIG // A1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9u
// SIG // ZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBvcmF0aW9u
// SIG // MTIwMAYDVQQDEylNaWNyb3NvZnQgUm9vdCBDZXJ0aWZp
// SIG // Y2F0ZSBBdXRob3JpdHkgMjAxMDAeFw0yMTA5MzAxODIy
// SIG // MjVaFw0zMDA5MzAxODMyMjVaMHwxCzAJBgNVBAYTAlVT
// SIG // MRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdS
// SIG // ZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9y
// SIG // YXRpb24xJjAkBgNVBAMTHU1pY3Jvc29mdCBUaW1lLVN0
// SIG // YW1wIFBDQSAyMDEwMIICIjANBgkqhkiG9w0BAQEFAAOC
// SIG // Ag8AMIICCgKCAgEA5OGmTOe0ciELeaLL1yR5vQ7VgtP9
// SIG // 7pwHB9KpbE51yMo1V/YBf2xK4OK9uT4XYDP/XE/HZveV
// SIG // U3Fa4n5KWv64NmeFRiMMtY0Tz3cywBAY6GB9alKDRLem
// SIG // jkZrBxTzxXb1hlDcwUTIcVxRMTegCjhuje3XD9gmU3w5
// SIG // YQJ6xKr9cmmvHaus9ja+NSZk2pg7uhp7M62AW36MEByd
// SIG // Uv626GIl3GoPz130/o5Tz9bshVZN7928jaTjkY+yOSxR
// SIG // nOlwaQ3KNi1wjjHINSi947SHJMPgyY9+tVSP3PoFVZht
// SIG // aDuaRr3tpK56KTesy+uDRedGbsoy1cCGMFxPLOJiss25
// SIG // 4o2I5JasAUq7vnGpF1tnYN74kpEeHT39IM9zfUGaRnXN
// SIG // xF803RKJ1v2lIH1+/NmeRd+2ci/bfV+AutuqfjbsNkz2
// SIG // K26oElHovwUDo9Fzpk03dJQcNIIP8BDyt0cY7afomXw/
// SIG // TNuvXsLz1dhzPUNOwTM5TI4CvEJoLhDqhFFG4tG9ahha
// SIG // YQFzymeiXtcodgLiMxhy16cg8ML6EgrXY28MyTZki1ug
// SIG // poMhXV8wdJGUlNi5UPkLiWHzNgY1GIRH29wb0f2y1BzF
// SIG // a/ZcUlFdEtsluq9QBXpsxREdcu+N+VLEhReTwDwV2xo3
// SIG // xwgVGD94q0W29R6HXtqPnhZyacaue7e3PmriLq0CAwEA
// SIG // AaOCAd0wggHZMBIGCSsGAQQBgjcVAQQFAgMBAAEwIwYJ
// SIG // KwYBBAGCNxUCBBYEFCqnUv5kxJq+gpE8RjUpzxD/LwTu
// SIG // MB0GA1UdDgQWBBSfpxVdAF5iXYP05dJlpxtTNRnpcjBc
// SIG // BgNVHSAEVTBTMFEGDCsGAQQBgjdMg30BATBBMD8GCCsG
// SIG // AQUFBwIBFjNodHRwOi8vd3d3Lm1pY3Jvc29mdC5jb20v
// SIG // cGtpb3BzL0RvY3MvUmVwb3NpdG9yeS5odG0wEwYDVR0l
// SIG // BAwwCgYIKwYBBQUHAwgwGQYJKwYBBAGCNxQCBAweCgBT
// SIG // AHUAYgBDAEEwCwYDVR0PBAQDAgGGMA8GA1UdEwEB/wQF
// SIG // MAMBAf8wHwYDVR0jBBgwFoAU1fZWy4/oolxiaNE9lJBb
// SIG // 186aGMQwVgYDVR0fBE8wTTBLoEmgR4ZFaHR0cDovL2Ny
// SIG // bC5taWNyb3NvZnQuY29tL3BraS9jcmwvcHJvZHVjdHMv
// SIG // TWljUm9vQ2VyQXV0XzIwMTAtMDYtMjMuY3JsMFoGCCsG
// SIG // AQUFBwEBBE4wTDBKBggrBgEFBQcwAoY+aHR0cDovL3d3
// SIG // dy5taWNyb3NvZnQuY29tL3BraS9jZXJ0cy9NaWNSb29D
// SIG // ZXJBdXRfMjAxMC0wNi0yMy5jcnQwDQYJKoZIhvcNAQEL
// SIG // BQADggIBAJ1VffwqreEsH2cBMSRb4Z5yS/ypb+pcFLY+
// SIG // TkdkeLEGk5c9MTO1OdfCcTY/2mRsfNB1OW27DzHkwo/7
// SIG // bNGhlBgi7ulmZzpTTd2YurYeeNg2LpypglYAA7AFvono
// SIG // aeC6Ce5732pvvinLbtg/SHUB2RjebYIM9W0jVOR4U3Uk
// SIG // V7ndn/OOPcbzaN9l9qRWqveVtihVJ9AkvUCgvxm2EhIR
// SIG // XT0n4ECWOKz3+SmJw7wXsFSFQrP8DJ6LGYnn8AtqgcKB
// SIG // GUIZUnWKNsIdw2FzLixre24/LAl4FOmRsqlb30mjdAy8
// SIG // 7JGA0j3mSj5mO0+7hvoyGtmW9I/2kQH2zsZ0/fZMcm8Q
// SIG // q3UwxTSwethQ/gpY3UA8x1RtnWN0SCyxTkctwRQEcb9k
// SIG // +SS+c23Kjgm9swFXSVRk2XPXfx5bRAGOWhmRaw2fpCjc
// SIG // ZxkoJLo4S5pu+yFUa2pFEUep8beuyOiJXk+d0tBMdrVX
// SIG // VAmxaQFEfnyhYWxz/gq77EFmPWn9y8FBSX5+k77L+Dvk
// SIG // txW/tM4+pTFRhLy/AsGConsXHRWJjXD+57XQKBqJC482
// SIG // 2rpM+Zv/Cuk0+CQ1ZyvgDbjmjJnW4SLq8CdCPSWU5nR0
// SIG // W2rRnj7tfqAxM328y+l7vzhwRNGQ8cirOoo6CGJ/2XBj
// SIG // U02N7oJtpQUQwXEGahC0HVUzWLOhcGbyoYIC2DCCAkEC
// SIG // AQEwggEAoYHYpIHVMIHSMQswCQYDVQQGEwJVUzETMBEG
// SIG // A1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9u
// SIG // ZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBvcmF0aW9u
// SIG // MS0wKwYDVQQLEyRNaWNyb3NvZnQgSXJlbGFuZCBPcGVy
// SIG // YXRpb25zIExpbWl0ZWQxJjAkBgNVBAsTHVRoYWxlcyBU
// SIG // U1MgRVNOOjE3OUUtNEJCMC04MjQ2MSUwIwYDVQQDExxN
// SIG // aWNyb3NvZnQgVGltZS1TdGFtcCBTZXJ2aWNloiMKAQEw
// SIG // BwYFKw4DAhoDFQCNMJ9r11RZj0PWu3uk+aQHF3IsVaCB
// SIG // gzCBgKR+MHwxCzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpX
// SIG // YXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4wHAYD
// SIG // VQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xJjAkBgNV
// SIG // BAMTHU1pY3Jvc29mdCBUaW1lLVN0YW1wIFBDQSAyMDEw
// SIG // MA0GCSqGSIb3DQEBBQUAAgUA53rtgzAiGA8yMDIzMDEy
// SIG // NTA3NTEzMVoYDzIwMjMwMTI2MDc1MTMxWjB4MD4GCisG
// SIG // AQQBhFkKBAExMDAuMAoCBQDneu2DAgEAMAsCAQACAwDI
// SIG // hQIB/zAHAgEAAgIRPjAKAgUA53w/AwIBADA2BgorBgEE
// SIG // AYRZCgQCMSgwJjAMBgorBgEEAYRZCgMCoAowCAIBAAID
// SIG // B6EgoQowCAIBAAIDAYagMA0GCSqGSIb3DQEBBQUAA4GB
// SIG // AIYMjJBzTMDoQDW7FeR7oqciDcqA5FakCvfRBa7rr4cF
// SIG // KxyRijDHTy+Uc8667zWhZiP2AsPcG+7dPcFIMdlOOjRo
// SIG // f4VW7GDWwEVEWgDDN9urcD+SZVgzfbWNLrefHOGAxhuP
// SIG // +8RpMNEAUYXPzWslMNxLHk8cb+FaESiAEAQIv7pCMYIE
// SIG // DTCCBAkCAQEwgZMwfDELMAkGA1UEBhMCVVMxEzARBgNV
// SIG // BAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQx
// SIG // HjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEm
// SIG // MCQGA1UEAxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAgUENB
// SIG // IDIwMTACEzMAAAG1rRrf14VwbRMAAQAAAbUwDQYJYIZI
// SIG // AWUDBAIBBQCgggFKMBoGCSqGSIb3DQEJAzENBgsqhkiG
// SIG // 9w0BCRABBDAvBgkqhkiG9w0BCQQxIgQgIVyQqlODcpNZ
// SIG // GmeJT2IBrMZcr6RR6n/kK2C8kB4z/HMwgfoGCyqGSIb3
// SIG // DQEJEAIvMYHqMIHnMIHkMIG9BCAnyg01LWhnFon2HNzl
// SIG // ZyKae2JJ9EvCXJVc65QIBfHIgzCBmDCBgKR+MHwxCzAJ
// SIG // BgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAw
// SIG // DgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3Nv
// SIG // ZnQgQ29ycG9yYXRpb24xJjAkBgNVBAMTHU1pY3Jvc29m
// SIG // dCBUaW1lLVN0YW1wIFBDQSAyMDEwAhMzAAABta0a39eF
// SIG // cG0TAAEAAAG1MCIEIHQqEhFkEFBWeQdbzCN3RKRnRdKb
// SIG // wfm70LQCmKbUTn3iMA0GCSqGSIb3DQEBCwUABIICACcb
// SIG // MCvT4icQ/Gac41Pa1osW6Vr5dclNr6eFqWyvjf/sXdf9
// SIG // djPtB32xx3Y/ZnbPlT7x0mbmFXRNQBEUxaayqQmosQM5
// SIG // 9pDwCxh2K2v3mrJZX2pybJRYS4y9vu4LMtmbOU8kKFjz
// SIG // 2J3q3Q9PBVXBn+HicNkTBV/Yr1/4nQE9f2ZUeYu+AGDm
// SIG // CcEurgLvBEsICIpSufHGyB8sIUMdtgxEOuGq8dnycg5s
// SIG // 8C8UZGCzbRXkGUICtBU6l26KhIKy4LmIBkLg3vI6ivbi
// SIG // Ls0cQBAqSwJNN/OuPTJ7uFoOpqDGn25AZpySgJ22Ehlj
// SIG // 1vubsrlnlETp9VnCE0aB22LeYWxRyrdJdjZGR0WsPXyd
// SIG // 6dYPG6geieLt3wz2mT0tUiEOv7GnHW7BsB6civ43/hdD
// SIG // lunwdpVSOaExE/gwSnWXEg1VZ6AqBGwEiYmUTzRLmukg
// SIG // Q66+r1rqkQO7NFOZdasrkxe+6WS4qk/elX1ETKP0/IJL
// SIG // xYoGEIZPBhz9ZXtCLPLHkhFKKm0m5IwSqBFPkr8Wv04a
// SIG // SHqI6A2r0GcT++K5uyb0GzJ1RA/Dq8QkBj5qcI26+Qw6
// SIG // +z6RAilj1ymw/cEzPDXwUONcwy/yg6PtoGgFnzGm/1sQ
// SIG // IL6YWqbVHuJm5FVi/ZWKABEi8ak0FPeEnfk+j4e2DqCP
// SIG // UPW1eg9Fu2cQFpwTfMly
// SIG // End signature block
